File: src/frontend/lexer.py
"""MiniLang lexical analyser.

Module M1a.  Owner: Member 1.
Status: COMPLETE (Week 2-3).

Hand-written scanner: no generator, no regex engine driving the token loop.
Longest-match is obtained by ordering the operator table by descending length;
keywords are recognised by looking an identifier up in KEYWORDS after it has
been scanned, which keeps the identifier rule a single loop.

Every token carries a (line, column) position so that every downstream
diagnostic can point at real source coordinates.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Iterator, List


class TokKind(Enum):
    # literals and names
    IDENT = auto()
    INT_LIT = auto()
    FLOAT_LIT = auto()
    # keywords
    KW_FN = auto()
    KW_LET = auto()
    KW_GLOBAL = auto()
    KW_IF = auto()
    KW_ELSE = auto()
    KW_WHILE = auto()
    KW_FOR = auto()
    KW_BREAK = auto()
    KW_CONTINUE = auto()
    KW_RETURN = auto()
    KW_TRUE = auto()
    KW_FALSE = auto()
    KW_INT = auto()
    KW_LONG = auto()
    KW_FLOAT = auto()
    KW_BOOL = auto()
    KW_VOID = auto()
    # operators and punctuation
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    ASSIGN = auto()
    EQ = auto()
    NE = auto()
    LT = auto()
    LE = auto()
    GT = auto()
    GE = auto()
    AND_AND = auto()
    OR_OR = auto()
    BANG = auto()
    AMP = auto()
    PIPE = auto()
    CARET = auto()
    TILDE = auto()
    SHL = auto()
    SHR = auto()
    ARROW = auto()
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    COMMA = auto()
    SEMI = auto()
    COLON = auto()
    EOF = auto()


KEYWORDS = {
    "fn": TokKind.KW_FN,
    "let": TokKind.KW_LET,
    "global": TokKind.KW_GLOBAL,
    "if": TokKind.KW_IF,
    "else": TokKind.KW_ELSE,
    "while": TokKind.KW_WHILE,
    "for": TokKind.KW_FOR,
    "break": TokKind.KW_BREAK,
    "continue": TokKind.KW_CONTINUE,
    "return": TokKind.KW_RETURN,
    "true": TokKind.KW_TRUE,
    "false": TokKind.KW_FALSE,
    "int": TokKind.KW_INT,
    "long": TokKind.KW_LONG,
    "float": TokKind.KW_FLOAT,
    "bool": TokKind.KW_BOOL,
    "void": TokKind.KW_VOID,
}

# Ordered longest-first so that '<<' wins over '<' and '->' over '-'.
OPERATORS = [
    ("&&", TokKind.AND_AND),
    ("||", TokKind.OR_OR),
    ("==", TokKind.EQ),
    ("!=", TokKind.NE),
    ("<=", TokKind.LE),
    (">=", TokKind.GE),
    ("<<", TokKind.SHL),
    (">>", TokKind.SHR),
    ("->", TokKind.ARROW),
    ("+", TokKind.PLUS),
    ("-", TokKind.MINUS),
    ("*", TokKind.STAR),
    ("/", TokKind.SLASH),
    ("%", TokKind.PERCENT),
    ("=", TokKind.ASSIGN),
    ("<", TokKind.LT),
    (">", TokKind.GT),
    ("!", TokKind.BANG),
    ("&", TokKind.AMP),
    ("|", TokKind.PIPE),
    ("^", TokKind.CARET),
    ("~", TokKind.TILDE),
    ("(", TokKind.LPAREN),
    (")", TokKind.RPAREN),
    ("{", TokKind.LBRACE),
    ("}", TokKind.RBRACE),
    ("[", TokKind.LBRACKET),
    ("]", TokKind.RBRACKET),
    (",", TokKind.COMMA),
    (";", TokKind.SEMI),
    (":", TokKind.COLON),
]


@dataclass(frozen=True)
class Token:
    kind: TokKind
    text: str
    line: int
    col: int

    def __repr__(self) -> str:  # pragma: no cover - debug aid
        return f"{self.kind.name}({self.text!r})@{self.line}:{self.col}"


class LexError(Exception):
    """Raised on an unrecoverable lexical error, carrying source position."""

    def __init__(self, message: str, line: int, col: int) -> None:
        super().__init__(f"{line}:{col}: lexical error: {message}")
        self.message = message
        self.line = line
        self.col = col


class Lexer:
    def __init__(self, src: str, filename: str = "<input>") -> None:
        self.src = src
        self.filename = filename
        self.i = 0
        self.line = 1
        self.col = 1

    # -- character helpers -------------------------------------------------
    def _peek(self, k: int = 0) -> str:
        j = self.i + k
        return self.src[j] if j < len(self.src) else ""

    def _advance(self) -> str:
        ch = self.src[self.i]
        self.i += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def _starts_with(self, s: str) -> bool:
        return self.src.startswith(s, self.i)

    @staticmethod
    def _is(ch: str, chars: str) -> bool:
        """Membership test that is False at end of input.

        Plain `ch in chars` is True when ch is the empty string, which sent
        every scanning loop past the end of the buffer.
        """
        return ch != "" and ch in chars

    # -- trivia ------------------------------------------------------------
    def _skip_trivia(self) -> None:
        while self.i < len(self.src):
            ch = self._peek()
            if self._is(ch, " \t\r\n"):
                self._advance()
            elif self._starts_with("//"):
                while self.i < len(self.src) and self._peek() != "\n":
                    self._advance()
            elif self._starts_with("/*"):
                start_line, start_col = self.line, self.col
                self._advance()
                self._advance()
                while True:
                    if self.i >= len(self.src):
                        raise LexError("unterminated block comment", start_line, start_col)
                    if self._starts_with("*/"):
                        self._advance()
                        self._advance()
                        break
                    self._advance()
            else:
                return

    # -- token rules -------------------------------------------------------
    def _lex_number(self) -> Token:
        line, col = self.line, self.col
        start = self.i
        if self._starts_with("0x") or self._starts_with("0X"):
            self._advance()
            self._advance()
            if not self._is(self._peek().lower(), "0123456789abcdef"):
                raise LexError("hexadecimal literal has no digits", line, col)
            while self._is(self._peek().lower(), "0123456789abcdef_"):
                self._advance()
            return Token(TokKind.INT_LIT, self.src[start:self.i], line, col)

        while self._peek().isdigit() or self._peek() == "_":
            self._advance()

        is_float = False
        # A '.' is part of the number only when followed by a digit, so that
        # '1..2' or a future member access does not silently become a float.
        if self._peek() == "." and self._peek(1).isdigit():
            is_float = True
            self._advance()
            while self._peek().isdigit() or self._peek() == "_":
                self._advance()
        if self._is(self._peek(), "eE") and (
            self._peek(1).isdigit()
            or (self._is(self._peek(1), "+-") and self._peek(2).isdigit())
        ):
            is_float = True
            self._advance()
            if self._is(self._peek(), "+-"):
                self._advance()
            while self._peek().isdigit():
                self._advance()

        kind = TokKind.FLOAT_LIT if is_float else TokKind.INT_LIT
        return Token(kind, self.src[start:self.i], line, col)

    def _lex_ident(self) -> Token:
        line, col = self.line, self.col
        start = self.i
        while self._peek().isalnum() or self._peek() == "_":
            self._advance()
        text = self.src[start:self.i]
        return Token(KEYWORDS.get(text, TokKind.IDENT), text, line, col)

    # -- driver ------------------------------------------------------------
    def tokens(self) -> Iterator[Token]:
        while True:
            self._skip_trivia()
            if self.i >= len(self.src):
                yield Token(TokKind.EOF, "", self.line, self.col)
                return

            ch = self._peek()
            if ch.isdigit():
                yield self._lex_number()
                continue
            if ch.isalpha() or ch == "_":
                yield self._lex_ident()
                continue

            for text, kind in OPERATORS:
                if self._starts_with(text):
                    line, col = self.line, self.col
                    for _ in text:
                        self._advance()
                    yield Token(kind, text, line, col)
                    break
            else:
                raise LexError(f"unexpected character {ch!r}", self.line, self.col)


def tokenize(src: str, filename: str = "<input>") -> List[Token]:
    """Convenience wrapper returning the full token list."""
    return list(Lexer(src, filename).tokens())

File: src/frontend/Lexer.cpp
#include "mtir/frontend/Lexer.h"

#include <cctype>
#include <cstddef>
#include <utility>

namespace mtir::ast {
namespace {

struct KindName {
  TokKind kind;
  std::string_view name;
};

// clang-format off
constexpr KindName kNames[] = {
  {TokKind::Ident, "IDENT"}, {TokKind::IntLit, "INT_LIT"}, {TokKind::FloatLit, "FLOAT_LIT"},
  {TokKind::KwFn, "KW_FN"}, {TokKind::KwLet, "KW_LET"}, {TokKind::KwGlobal, "KW_GLOBAL"},
  {TokKind::KwIf, "KW_IF"}, {TokKind::KwElse, "KW_ELSE"}, {TokKind::KwWhile, "KW_WHILE"},
  {TokKind::KwFor, "KW_FOR"}, {TokKind::KwBreak, "KW_BREAK"},
  {TokKind::KwContinue, "KW_CONTINUE"}, {TokKind::KwReturn, "KW_RETURN"},
  {TokKind::KwTrue, "KW_TRUE"}, {TokKind::KwFalse, "KW_FALSE"},
  {TokKind::KwInt, "KW_INT"}, {TokKind::KwLong, "KW_LONG"}, {TokKind::KwFloat, "KW_FLOAT"},
  {TokKind::KwBool, "KW_BOOL"}, {TokKind::KwVoid, "KW_VOID"},
  {TokKind::Plus, "PLUS"}, {TokKind::Minus, "MINUS"}, {TokKind::Star, "STAR"},
  {TokKind::Slash, "SLASH"}, {TokKind::Percent, "PERCENT"},
  {TokKind::Assign, "ASSIGN"}, {TokKind::Eq, "EQ"}, {TokKind::Ne, "NE"},
  {TokKind::Lt, "LT"}, {TokKind::Le, "LE"}, {TokKind::Gt, "GT"}, {TokKind::Ge, "GE"},
  {TokKind::AndAnd, "AND_AND"}, {TokKind::OrOr, "OR_OR"}, {TokKind::Bang, "BANG"},
  {TokKind::Amp, "AMP"}, {TokKind::Pipe, "PIPE"}, {TokKind::Caret, "CARET"},
  {TokKind::Tilde, "TILDE"}, {TokKind::Shl, "SHL"}, {TokKind::Shr, "SHR"},
  {TokKind::Arrow, "ARROW"},
  {TokKind::LParen, "LPAREN"}, {TokKind::RParen, "RPAREN"},
  {TokKind::LBrace, "LBRACE"}, {TokKind::RBrace, "RBRACE"},
  {TokKind::LBracket, "LBRACKET"}, {TokKind::RBracket, "RBRACKET"},
  {TokKind::Comma, "COMMA"}, {TokKind::Semi, "SEMI"}, {TokKind::Colon, "COLON"},
  {TokKind::Eof, "EOF"},
};
// clang-format on

} // namespace

std::string_view tokKindName(TokKind kind) {
  for (const KindName &entry : kNames)
    if (entry.kind == kind)
      return entry.name;
  return "<invalid>";
}

bool isTypeKeyword(TokKind kind) {
  return kind == TokKind::KwInt || kind == TokKind::KwLong ||
         kind == TokKind::KwFloat || kind == TokKind::KwBool ||
         kind == TokKind::KwVoid;
}

} // namespace mtir::ast

namespace mtir::frontend {
namespace {

using ast::TokKind;
using ast::Token;

struct Keyword {
  std::string_view text;
  TokKind kind;
};

constexpr Keyword kKeywords[] = {
    {"fn", TokKind::KwFn}, {"let", TokKind::KwLet}, {"global", TokKind::KwGlobal},
    {"if", TokKind::KwIf}, {"else", TokKind::KwElse}, {"while", TokKind::KwWhile},
    {"for", TokKind::KwFor}, {"break", TokKind::KwBreak},
    {"continue", TokKind::KwContinue}, {"return", TokKind::KwReturn},
    {"true", TokKind::KwTrue}, {"false", TokKind::KwFalse},
    {"int", TokKind::KwInt}, {"long", TokKind::KwLong}, {"float", TokKind::KwFloat},
    {"bool", TokKind::KwBool}, {"void", TokKind::KwVoid},
};

/// Ordered longest-first, so that '<<' wins over '<' and '->' over '-'.
constexpr Keyword kOperators[] = {
    {"&&", TokKind::AndAnd}, {"||", TokKind::OrOr}, {"==", TokKind::Eq},
    {"!=", TokKind::Ne}, {"<=", TokKind::Le}, {">=", TokKind::Ge},
    {"<<", TokKind::Shl}, {">>", TokKind::Shr}, {"->", TokKind::Arrow},
    {"+", TokKind::Plus}, {"-", TokKind::Minus}, {"*", TokKind::Star},
    {"/", TokKind::Slash}, {"%", TokKind::Percent}, {"=", TokKind::Assign},
    {"<", TokKind::Lt}, {">", TokKind::Gt}, {"!", TokKind::Bang},
    {"&", TokKind::Amp}, {"|", TokKind::Pipe}, {"^", TokKind::Caret},
    {"~", TokKind::Tilde}, {"(", TokKind::LParen}, {")", TokKind::RParen},
    {"{", TokKind::LBrace}, {"}", TokKind::RBrace}, {"[", TokKind::LBracket},
    {"]", TokKind::RBracket}, {",", TokKind::Comma}, {";", TokKind::Semi},
    {":", TokKind::Colon},
};

bool isDigit(char c) { return std::isdigit(static_cast<unsigned char>(c)) != 0; }
bool isAlpha(char c) { return std::isalpha(static_cast<unsigned char>(c)) != 0; }
bool isAlnum(char c) { return std::isalnum(static_cast<unsigned char>(c)) != 0; }

bool isHexDigit(char c) {
  return isDigit(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}

class Lexer {
public:
  Lexer(std::string_view source, std::string file)
      : src_(source), file_(std::move(file)) {}

  LexResult run();

private:
  char peek(std::size_t k = 0) const {
    const std::size_t j = i_ + k;
    return j < src_.size() ? src_[j] : '\0';
  }

  char advance() {
    const char c = src_[i_++];
    if (c == '\n') {
      ++line_;
      col_ = 1;
    } else {
      ++col_;
    }
    return c;
  }

  bool startsWith(std::string_view s) const { return src_.compare(i_, s.size(), s) == 0; }

  support::SourceLoc here() const { return support::SourceLoc{line_, col_}; }

  void error(support::SourceLoc loc, std::string message) {
    support::Location l;
    l.file = file_;
    l.loc = loc;
    diags_.push_back(support::error("E000", std::move(message), l));
  }

  void skipTrivia();
  Token lexNumber();
  Token lexIdent();

  std::string_view src_;
  std::string file_;
  std::size_t i_ = 0;
  int line_ = 1;
  int col_ = 1;
  support::Diagnostics diags_;
};

void Lexer::skipTrivia() {
  while (i_ < src_.size()) {
    const char c = peek();
    if (c == ' ' || c == '\t' || c == '\r' || c == '\n') {
      advance();
    } else if (startsWith("//")) {
      while (i_ < src_.size() && peek() != '\n')
        advance();
    } else if (startsWith("/*")) {
      const support::SourceLoc start = here();
      advance();
      advance();
      for (;;) {
        if (i_ >= src_.size()) {
          error(start, "unterminated block comment");
          return;
        }
        if (startsWith("*/")) {
          advance();
          advance();
          break;
        }
        advance();
      }
    } else {
      return;
    }
  }
}

Token Lexer::lexNumber() {
  const support::SourceLoc loc = here();
  const std::size_t start = i_;

  if (startsWith("0x") || startsWith("0X")) {
    advance();
    advance();
    if (!isHexDigit(peek())) {
      error(loc, "hexadecimal literal has no digits");
      // Still produce a token so the parser sees something plausible.
      return Token{TokKind::IntLit, std::string(src_.substr(start, i_ - start)), loc};
    }
    while (isHexDigit(peek()) || peek() == '_')
      advance();
    return Token{TokKind::IntLit, std::string(src_.substr(start, i_ - start)), loc};
  }

  while (isDigit(peek()) || peek() == '_')
    advance();

  bool isFloat = false;
  // A '.' is part of the number only when a digit follows, so "1..2" does not
  // silently become a float (docs/minilang-spec.md section 1).
  if (peek() == '.' && isDigit(peek(1))) {
    isFloat = true;
    advance();
    while (isDigit(peek()) || peek() == '_')
      advance();
  }
  if ((peek() == 'e' || peek() == 'E') &&
      (isDigit(peek(1)) || ((peek(1) == '+' || peek(1) == '-') && isDigit(peek(2))))) {
    isFloat = true;
    advance();
    if (peek() == '+' || peek() == '-')
      advance();
    while (isDigit(peek()))
      advance();
  }

  return Token{isFloat ? TokKind::FloatLit : TokKind::IntLit,
               std::string(src_.substr(start, i_ - start)), loc};
}

Token Lexer::lexIdent() {
  const support::SourceLoc loc = here();
  const std::size_t start = i_;
  while (isAlnum(peek()) || peek() == '_')
    advance();
  const std::string text(src_.substr(start, i_ - start));
  for (const Keyword &kw : kKeywords)
    if (kw.text == text)
      return Token{kw.kind, text, loc};
  return Token{TokKind::Ident, text, loc};
}

LexResult Lexer::run() {
  LexResult result;
  for (;;) {
    skipTrivia();
    if (i_ >= src_.size()) {
      result.tokens.push_back(Token{TokKind::Eof, "", here()});
      break;
    }

    const char c = peek();
    if (isDigit(c)) {
      result.tokens.push_back(lexNumber());
      continue;
    }
    if (isAlpha(c) || c == '_') {
      result.tokens.push_back(lexIdent());
      continue;
    }

    bool matched = false;
    for (const Keyword &op : kOperators) {
      if (startsWith(op.text)) {
        const support::SourceLoc loc = here();
        for (std::size_t k = 0; k < op.text.size(); ++k)
          advance();
        result.tokens.push_back(Token{op.kind, std::string(op.text), loc});
        matched = true;
        break;
      }
    }
    if (!matched) {
      const support::SourceLoc loc = here();
      error(loc, "unexpected character '" + std::string(1, c) + "'");
      // Skip it and keep going, so one bad byte does not hide the rest.
      advance();
    }
  }
  result.diagnostics = std::move(diags_);
  return result;
}

} // namespace

LexResult tokenize(std::string_view source, std::string file) {
  return Lexer(source, std::move(file)).run();
}

} // namespace mtir::frontend
